sympy/solvers/inequalities.py

```diff
@@ -560,21 +543,40 @@ def valid(x):
                         'relationship did not evaluate: %s' % r)
 
             singularities = []
-            for d in denoms(expr, gen):
-                singularities.extend(solvify(d, gen, domain))
-            if not continuous:
-                domain = continuous_domain(expanded_e, gen, domain)
 
             include_x = '=' in expr.rel_op and expr.rel_op != '!='
+            try:
+                for d in denoms(expr, gen):
+                    singularities.extend(solvify(d, gen, domain))
+                if not continuous:
+                    domain = continuous_domain(expanded_e, gen, domain)
+                if gen not in n.free_symbols and len(e.free_symbols) > 1:
+                    raise ValueError
+                # this might raise ValueError on its own
+                # or it might give None...
+                solns = solvify(e, gen, domain)
+                if solns is None:
+                    # in which case we raise ValueError
+                    raise ValueError
+                else:
+                    discontinuities = set(domain.boundary -
+                        FiniteSet(domain.inf, domain.sup))
+                    # remove points that are not between inf and sup of domain
+                    critical_points = FiniteSet(*(solns + singularities + list(
```

Comment:
Actually, the `critical_points ` are converted to `Intersection` due to the `solns`(which are o/p of `solvify` see line L-557 ) which is the reason for `TypeError`,...